Use composition instead of inheritence for configuring bindings

Dangles91 opened the pull request:
This change introduces BindingContext as an instance class to create bindings against and register for notification events.

This simplifies the implementation of IObservableObject and adds flexibility by removing the requirement to inherit from IObservableNode. 

Objects on the binding path must still implement IObservableObject to benefit from binding, and the root node passes to the constructor of BindingContext must also implement IObservableObject.

Several basic classes exist that can be inherited from to ease initial development; ControlViewModel, NodeViewModel, and ObservableObject.

BindingContext does not ineherit from Godot.Node, so bindings require a reference to the scene tree node/control and not just a path.

Dangles91 commented on Binding/Binding.cs:
Register for PropertyChanged events if the target is IObservableObject - most likely a binding for a custom control or node

avilches commented on Binding/ViewModel.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Godot.Community.ControlBinding;
+
+public abstract partial class NodeViewModel : Node, IViewModel
```
Maybe you can implement something similar to: https://github.com/canton7/PropertyChanged.SourceGenerator 
So users will not need to extend from this two classes and they will not need to create the boilerplate  :)
```C#
private bool labelIsVisible = true;
    public bool IsAddNewPlayerEnabled
    {
        get { return labelIsVisible; }
        set { this.SetValue(ref labelIsVisible, value); }
    }
```

Dangles91 replied:
It might even be possible for users to use it directly if we use INotifyPropertyChanged instead of a custom event handler. 

The custom event handler was originally required to support Godot Signals in C#, but after moving to raw events, we're no longer restricted.

Dangles91 commented:
@avilches - see the latest commits here in response to adding codegen. After switching to INotifyPropertyChanged, it is essentially plug and play and also works with the basic inheritable classes if anyone starts without codegen and decides to migrate.

sseles commented:
Oh, this is great; I was just having problems dealing with my control. This really helps out. Thank you!

Dangles91 commented:
> Oh, this is great; I was just having problems dealing with my control. This really helps out. Thank you!

No problem, that's actually great to hear! 

I'm just working on implementing INotifyCollectionChanged instead of using the custom IObservableList implementation and then this should be ready for a release.

Dangles91 commented:
### Breaking Changes

- Bindings now use System.ComponentModel INotifyPropertyChanged and INotifyCollectionChanged
- Removed ObservableList. Use ObservableCollection instead
- Scene items no longer inherit from ObservableNode
- Scene items still have to implement IObservableObject
- Scene items must create an instance of BindingContext and register bindings against the context
- BindingContext is now the context for validations. Use this for HasError and for subscribing to the ControlValidationChanged event
- New base classes ControlViewModel and NodeViewModel can be used isntead of implementing IObservableObject
- Bindings require a node reference instead of a path

### New Features

- Code generators such as PropertyChanged.SourceGenerator can be used to implement INotifyPropertyChanged
- List bindings now support Move, Replace, and Insert events
- Added support for custom control bindings. Custom controls must implement IObservableObject
- Custom controls can be bound by targeting a property that raises the OnPropertyChanged event
- Nodes can be bound using extensions methods for BindProperty etc.

cc @sseles - I'll be doing some testing for this in another project using the pre-release package. Feel free to check it out. I'll update the docs to address all of the above changes for the stable release
